Document the Connect component API (#14931)
PipedreamHQ/pipedream: docs-v2/pages/connect/components.mdx

import { Steps, Tabs } from 'nextra/components'
import ArcadeEmbed from '@/components/ArcadeEmbed'
import Callout from '@/components/Callout'
import Image from 'next/image'

> # Running Components for Your End Users

Reviewer: ```suggestion
# Running components for your end users
```

(see https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings)

Pipedream Connect provides you with the ability to pre-configure, deploy and
invoke [components](/components) **on behalf of [your end
users](/connect/api#external-users)**.

> ## What are Components?

Reviewer: Same sentence-case suggestion here — I won't add comments for the remainder but if you can update the rest of the headings that would help

In the context of Pipedream Connect, components are self-contained executable
units of code. They are usually provided with some arbitrary input and produce a
result that's exported as output. These components are developed by the
Pipedream community and their source code is available in our [public Github
repository](https://github.com/PipedreamHQ/pipedream/tree/master/components).

To learn more about components in detail, please visit our [Components
guide](/components).

> ## How to run components for your end users

Reviewer: Do you see how `pages/connect/workflows.mdx` uses the `Steps` component? You should be able to wrap the headers in this section in `<Steps> ... </Steps>` and it will generate the nice numbered sections on build.

### Retrieving a Component's Definition

The first thing to do in order to run a component for your end users is to know
about the component itself. To start with, you need to know the component's
unique key. Using that key, you can retrieve the component's structure from the
Pipedream API. See the [component
structure](/components/api#component-structure) section in our docs for more
details.

> As an example, the following API call will return the structure of the "List
> Commits" component for Gitlab:

Reviewer: ```suggestion
As an example, the following API call will return the structure of the **List
Commits** component for Gitlab:
```

```text
GET /v1/connect/components/gitlab-list-commits
```

The response will contain the component's structure, including its
(human-understandable) name, version, and most importantly, the configuration

> options that the component accepts (also known as "props", which is an
> abbreviation of "properties"). Here's an example of the response for the

Reviewer: ```suggestion
options that the component accepts (also known as [props](/components/api#props) or "properties"). Here's an example of the response for the
```

component in the example above:

```json
{
  "data": {
    "name": "List Commits",
    "version": "0.0.3",
    "key": "gitlab-list-commits",
    "configurable_props": [
      {
        "name": "gitlab",
        "type": "app",
        "app": "gitlab"
      },
      {
        "name": "projectId",
        "type": "integer",
        "label": "Project ID",
        "description": "The project ID, as displayed in the main project page",
        "remoteOptions": true
      },
      {
        "name": "refName",
        "type": "string",
        "label": "Branch Name",
        "description": "The name of the branch",
        "remoteOptions": true,
        "optional": true
      },
      {
        "name": "max",
        "type": "integer",
        "label": "Max Results",
        "description": "Max number of results to return. Default value is `100`",
        "optional": true,
        "default": 100
      }
    ]
  }
}
```

Using this information, you can now drive the configuration of the component for
your end users, as described in the next section.

### Configuration

Component execution on behalf of your end users requires a few preliminary

> steps, focused on getting the right input parameters (a.k.a.

Reviewer: Most style guides prefer just "aka" with no periods

[props](/workflows/using-props)) to the component.

Configuring each prop for a component usually involves an API call to our
backend to retrieve the possible values, unless the values that a prop can take
are static/free-form. The endpoint is accessible at:

```text
POST /v1/connect/components/configure
```

> The reason for this back-and-forth is because the options for a prop are usually
> linked to a specific user's account, and the Pipedream API is the one in charge
> of contacting the third-party service using the user's credentials, retrieving
> the necessary data, and sending these options back as a response. Please not
> that this logic is not implemented by the API itself, but by the component's
> code; in particular, the `options` function of the corresponding prop.

Reviewer: ```suggestion
Typically, the options for a prop are linked to a specific user's account. Each of these props implements
an `options` method that retrieves the necessary options from the third-party API, formats them, and sends them back in the response for the end user to select.
```

Reviewer: (just some advice for things to make it shorter)

> The initial configuration call must contain your user's account ID for the

Reviewer: This payload contains a lot more than just the `authProvisionId` and `prop_name`, so as a naive user I'm a little confused when I see more params, especially things like `async_handle` and `external_user_id`. Maybe describe those params and change this to a list to walk people through each param, e.g.

> When you call the `/v1/connect/components/configure` endpoint, you must pass:
>  * The `id` of the component — the component's `key`
> * The [`external_user_id`](/connect/api#external-users)
> ...

> target app, and the name of the prop you want to configure. Using the ["List
> Commits" component for

Reviewer: ```suggestion
target app, and the name of the prop you want to configure. Using the [**List
Commits** component for
```

Gitlab](https://github.com/PipedreamHQ/pipedream/blob/master/components/gitlab/actions/list-commits/list-commits.mjs#L4)
as an example, the payload sent to the configuration API would look like this:

```json
{
  "async_handle": "qxyTp7QcS2A6",
  "external_user_id": "demo-34c13d13-a31e-4a3d-8b63-0ac954671095",
  "id": "gitlab-list-commits",
  "prop_name": "projectId",
  "configured_props": {
    "gitlab": {
      "authProvisionId": "apn_oOhaBlD"
    }
  }
}
```

The goal of this particular call is to retrieve the options for the `projectId`
prop of the "List Commits" component for Gitlab, for the user with the ID
`demo-34c13d13-a31e-4a3d-8b63-0ac954671095`. The `authProvisionId` is the ID of
the account that the user has connected to the Gitlab app (see [this
section](workflows#configure-accounts-to-use-your-end-users-auth) for more

> details on how to create these accounts).

Reviewer: This should probably just be combined with the section above if you're documenting all params there.

```json
{
  "observations": [],
  "context": null,
  "options": [
    {
      "label": "jverce/foo-massive-10231-1",
      "value": 45672541
    },
    {
      "label": "jverce/foo-massive-10231",
      "value": 45672514
    },
    {
      "label": "jverce/foo-massive-14999-2",
      "value": 45672407
    },
    {
      "label": "jverce/foo-massive-14999-1",
      "value": 45672382
    },
    {
      "label": "jverce/foo-massive-14999",
      "value": 45672215
    },
    {
      "label": "jverce/gitlab-development-kit",
      "value": 21220953
    },
    {
      "label": "jverce/gitlab",
      "value": 21208123
    }
  ],
  "errors": [],
  "timings": {
    "api_to_sidekiq": 1734043172355.1042,
    "sidekiq_received": 1734043172357.867,
    "sidekiq_to_lambda": 1734043172363.6812,
    "sidekiq_done": 1734043173461.6406,
    "lambda_configure_prop_called": 1734043172462,
    "lambda_done": 1734043173455
  },
  "stringOptions": null
}
```

<Callout type="info">

Fields inside `configured_props` are written in camel case since they refer to
the names of props as they appear in the component's code, they are not
attributes that the API itself consumes. For example, to retrieve the

> configuration options for the `refName` prop:

Reviewer: Did you intend to add code here?

Author: A copy/paste gone wrong 😅

</Callout>

> Subsequent prop configuration calls will be identical to the one above, but will
> incrementally add more props to the `configured_props` object, with the name of
> the prop being configured as the value for the `prop_name` attribute.

Reviewer: ```suggestion
You configure props one-by-one, making a call to the component configuration API for each new prop. Subsequent prop configuration calls will be identical to the one above:

1. Add the prop you currently want to configure as the `prop_name`
2. Include the names and values of all previously-configured props in the `configured_props` object. Keep this object in your app's local state, add a prop once you or the end user selects a value, and pass it to the `configured_props` API param.
```
